File: app/src/main/java/com/ubiqsmart/ui/main/adapter/WalletAdapter.java
package com.ubiqsmart.ui.main.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.ubiqsmart.repository.data.TransactionDisplay;
import com.ubiqsmart.utils.AddressNameConverter;
import com.ubiqsmart.utils.Blockies;
import com.ubiqsmart.utils.ExchangeCalculator;
import me.grantland.widget.AutofitTextView;
import com.ubiqsmart.R;
import com.ubiqsmart.repository.data.Wallet;

import java.util.*;

public class WalletAdapter extends RecyclerView.Adapter<WalletAdapter.MyViewHolder> {

  private Context context;
  private List<Wallet> boxlist;
  private int lastPosition = -1;
  private View.OnClickListener listener;
  private View.OnCreateContextMenuListener contextMenuListener;
  private int position;

  public static class MyViewHolder extends RecyclerView.ViewHolder {

    public TextView walletname, walletbalance;
    public ImageView addressimage, type;
    AutofitTextView walletaddress;
    public LinearLayout container;

    public MyViewHolder(View view) {
      super(view);

      walletaddress = view.findViewById(R.id.walletaddress);
      walletname = view.findViewById(R.id.wallet_name);
      walletbalance = view.findViewById(R.id.wallet_balance);
      addressimage = view.findViewById(R.id.addressimage);
      type = view.findViewById(R.id.type);
      container = view.findViewById(R.id.container);
    }

    public void clearAnimation() {
      container.clearAnimation();
    }
  }

  public WalletAdapter(List<Wallet> boxlist, Context context, View.OnClickListener listener, View.OnCreateContextMenuListener l) {
    this.boxlist = boxlist;
    this.context = context;
    this.listener = listener;
    this.contextMenuListener = l;
  }

  @Override public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_w_address, parent, false);

    itemView.setOnClickListener(listener);
    itemView.setOnCreateContextMenuListener(contextMenuListener);
    return new MyViewHolder(itemView);
  }

  @Override public void onBindViewHolder(MyViewHolder holder, final int position) {
    Wallet box = boxlist.get(position);

    holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
      @Override public boolean onLongClick(View v) {
        setPosition(position);
        return false;
      }
    });
    holder.walletaddress.setText(box.getPublicKey());
    String walletname = AddressNameConverter.getInstance(context).get(box.getPublicKey());
    holder.walletname.setText(walletname == null ? "New Wallet" : walletname);
    if (box.getType() != Wallet.CONTACT) {
      holder.walletbalance.setText(ExchangeCalculator.getInstance()
          .displayBalanceNicely(
              ExchangeCalculator.getInstance().convertRate(box.getBalance(), ExchangeCalculator.getInstance().getCurrent().getRate()))
          + " "
          + ExchangeCalculator.getInstance().getCurrencyShort());
    }
    holder.addressimage.setImageBitmap(Blockies.createIcon(box.getPublicKey()));

    holder.type.setVisibility(box.getType() == TransactionDisplay.NORMAL || box.getType() == Wallet.CONTACT ? View.INVISIBLE : View.VISIBLE);

    setAnimation(holder.container, position);
  }

  public int getPosition() {
    return position;
  }

  public void setPosition(int position) {
    this.position = position;
  }

  @Override public void onViewRecycled(WalletAdapter.MyViewHolder holder) {
    holder.itemView.setOnLongClickListener(null);
    super.onViewRecycled(holder);
  }

  private void setAnimation(View viewToAnimate, int position) {
    if (position > lastPosition) {
      Animation animation = AnimationUtils.loadAnimation(context, (position > lastPosition) ? R.anim.up_from_bottom : R.anim.down_from_bottom);
      viewToAnimate.startAnimation(animation);
      lastPosition = position;
    }
  }

  @Override public void onViewDetachedFromWindow(MyViewHolder holder) {
    holder.clearAnimation();
  }

  @Override public int getItemCount() {
    return boxlist.size();
  }
}

File: app/src/main/java/com/ubiqsmart/app/ui/main/adapter/WalletAdapter.kt
package com.ubiqsmart.app.ui.main.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.ubiqsmart.R
import com.ubiqsmart.app.utils.AddressNameConverter
import com.ubiqsmart.app.utils.Blockies
import com.ubiqsmart.app.utils.ExchangeCalculator
import com.ubiqsmart.domain.models.TransactionDisplay
import com.ubiqsmart.domain.models.WalletEntry
import me.grantland.widget.AutofitTextView

class WalletAdapter(
    private val context: Context,
    private val wallets: List<WalletEntry>,
    private val listener: View.OnClickListener,
    private val contextMenuListener: View.OnCreateContextMenuListener
) : RecyclerView.Adapter<com.ubiqsmart.app.ui.main.adapter.WalletAdapter.WalletViewHolder>() {

  private var lastPosition = -1
  var position = 0

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WalletViewHolder {
    val itemView = LayoutInflater.from(parent.context).inflate(R.layout.wallet_item_adapter, parent, false)
    itemView.setOnClickListener(listener)
    itemView.setOnCreateContextMenuListener(contextMenuListener)
    return WalletViewHolder(itemView)
  }

  override fun onBindViewHolder(holder: WalletViewHolder, position: Int) {
    val box = wallets[position]

    val exchangeCalculator = ExchangeCalculator.getInstance()
    val walletName = AddressNameConverter.getInstance(context).get(box.publicKey)

    holder.apply {
      walletaddress.text = box.publicKey
      walletname.text = walletName ?: context.getString(R.string.new_wallet)
      if (box.type != WalletEntry.CONTACT) {
        holder.walletbalance.text = String.format("%s %s",
            exchangeCalculator.displayBalanceNicely(exchangeCalculator.convertRate(box.balance, exchangeCalculator.current.rate)),
            exchangeCalculator.currencyShort)
      }
      addressimage.setImageBitmap(Blockies.createIcon(box.publicKey))
      type.visibility =
          if (box.type == TransactionDisplay.NORMAL || box.type == WalletEntry.CONTACT) {
            View.INVISIBLE
          } else {
            View.VISIBLE
          }
    }

    holder.itemView.setOnLongClickListener {
      this.position = position
      false
    }

    setAnimation(holder.container, position)
  }

  override fun onViewRecycled(holder: WalletViewHolder?) {
    holder!!.itemView.setOnLongClickListener(null)
    super.onViewRecycled(holder)
  }

  override fun onViewDetachedFromWindow(holder: WalletViewHolder?) {
    holder!!.clearAnimation()
  }

  override fun getItemCount() = wallets.size

  private fun setAnimation(viewToAnimate: View, position: Int) {
    if (position > lastPosition) {
      val animation = AnimationUtils.loadAnimation(context, if (position > lastPosition) R.anim.up_from_bottom else R.anim.down_from_bottom)
      viewToAnimate.startAnimation(animation)
      lastPosition = position
    }
  }

  class WalletViewHolder(view: View) : RecyclerView.ViewHolder(view) {

    var walletname: TextView = view.findViewById(R.id.wallet_name)
    var walletbalance: TextView = view.findViewById(R.id.wallet_balance)
    var addressimage: ImageView = view.findViewById(R.id.address_image_view)
    var type: ImageView = view.findViewById(R.id.type)
    var walletaddress: AutofitTextView = view.findViewById(R.id.walletaddress)
    var container: LinearLayout = view.findViewById(R.id.container)

    fun clearAnimation() {
      container.clearAnimation()
    }
  }
}
